R/score.R

```diff
@@ -111,52 +107,38 @@
 #' @export
 
 score <- function(data,
-                  metrics = NULL,
                   ...) {
 
+  # need to deal with the fact that the argument is now called "x" instead of "data".
+  # can we get around that and keep calling it data?
+
   # preparations ---------------------------------------------------------------
   if (is_scoringutils_check(data)) {
     check_data <- data
   } else {
     check_data <- check_forecasts(data)
   }
 
-  data <- check_data$cleaned_data
+  score_data <- copy(check_data$cleaned_data)
   prediction_type <- check_data$prediction_type
   forecast_unit <- check_data$forecast_unit
   target_type <- check_data$target_type
 
   # check metrics are available or set to all metrics --------------------------
-  metrics <- check_metrics(metrics)
+  setattr(score_data, "forecast_unit", forecast_unit)
+  setattr(score_data, "target_type", target_type)
+  setattr(score_data, "prediction_type", prediction_type)
 
-  # Score binary predictions ---------------------------------------------------
-  if (target_type == "binary") {
-    scores <- score_binary(
-      data = data,
-      forecast_unit = forecast_unit,
-      metrics = metrics
-    )
-  }
+  class(score_data) <- c(class(score_data), "scoringutils")
 
-  # Score quantile predictions -------------------------------------------------
-  if (prediction_type == "quantile") {
-    scores <- score_quantile(
-      data = data,
-      forecast_unit = forecast_unit,
-      metrics = metrics,
-      ...
-    )
-  }
-
-  # Score integer or continuous predictions ------------------------------------
-  if (prediction_type %in% c("integer", "continuous") && (target_type != "binary")) {
-    scores <- score_sample(
-      data = data,
-      forecast_unit = forecast_unit,
-      metrics = metrics,
-      prediction_type = prediction_type
-    )
+  # do some kind of checking for which attribute is appropriate.
+  if (target_type == "binary") {
+    class(score_data) <- c("scoringutils_binary", class(score_data))
+  } else if (prediction_type == "quantile") {
+    class(score_data) <- c("scoringutils_quantile", class(score_data))
+  } else if (prediction_type %in% c("integer", "continuous")) {
+    class(score_data) <- c("scoringutils_sample", class(score_data))
   }
 
-  return(scores[])
+  UseMethod(generic = "score", object = score_data)
```

Comment:
I would have thought we would have score, score.default and then the specific methods where score.default calls check forecast and then calls itself (score)? Then make check_forecast a true class constructor (above code looks like it just reorganises?)